warehouse/query-core/src/main/java/datawave/query/jexl/functions/GroupingRequiredFilterFunctions.java

```diff
@@ -211,122 +242,138 @@
-                        matches.addAll(EvaluationPhaseFilterFunctions.includeRegex(fv, regex));
-                        matches.add(match);
-                    }
-                    
+                    Object fieldValue = args[i];
+                    String fieldName = ValueTuple.getFieldName(fieldValue);
+                    // @formatter:off
+                    manageMatchesInGroupLeftRemainingArgs(fieldValue,
+                            args[i + 1].toString(), // regex
+                            allMatches, theFirstMatch,
+                            EvaluationPhaseFilterFunctions.getMatchToLeftOfPeriod(fieldName, positionFromLeft), // the next match
+                            currentMatch);
+                    // @formatter:on
                 }
             }
         }
         
         // if there was a match found at all levels, then the matches.size will be equal to the
         // number of field/regex pairs
-        if (matches.size() < args.length / 2) { // truncated in case args.length was odd
-            matches.clear();
+        if (allMatches.size() < args.length / 2) { // truncated in case args.length was odd
+            allMatches.clear();
+        }
+        if (log.isTraceEnabled()) {
+            log.trace("returning matches:" + allMatches);
+        }
+        return Collections.unmodifiableCollection(allMatches);
+    }
+    
+    private static void manageMatchesInGroupLeftRemainingArgs(Object fieldValue, String regex, Collection<ValueTuple> allMatches, String theFirstMatch,
+                    String theNextMatch, ValueTuple currentMatch) {
+        
+        if (theNextMatch != null && theNextMatch.equals(theFirstMatch)) {
+            if (log.isTraceEnabled()) {
+                log.trace("\tfirst match equals the second: " + theFirstMatch + " == " + theNextMatch);
+            }
+            allMatches.addAll(EvaluationPhaseFilterFunctions.includeRegex(fieldValue, regex));
+            allMatches.add(currentMatch);
         }
-        log.trace("returning matches:" + matches);
-        return Collections.unmodifiableCollection(matches);
     }
     
-    public static Collection<ValueTuple> atomValuesMatch(Object... args) {
+    /**
+     * test that fields (field names) have values that match within the same grouping context.
+     * 
+     * @param fields
+     * @return a collection of matches
+     * */
+    public static Collection<ValueTuple> atomValuesMatch(Object... fields) {
         List<Iterable<?>> iterableFields = new ArrayList<>();
         
-        for (Object field : args) {
+        for (Object field : fields) {
             if (field instanceof Iterable == false) {
                 field = Collections.singleton(field);
             }
             iterableFields.add((Iterable<?>) field);
         }
-        return atomValuesMatch(iterableFields);
+        return atomValuesMatch(iterableFields.toArray(new Iterable[iterableFields.size()]));
```

Comment:
IntelliJ tells me that `iterbleFields.toArray(new Iterable[0])` is faster and safer here. Here's another article on the subject: https://medium.com/@hanru.yeh/collection-toarray-new-t-0-will-be-faster-than-collection-toarray-new-t-size-5b129fabfa88